osu.Game.Tournament/Models/TournamentMatch.cs

```
@@ -118,6 +121,56 @@ public void StartMatch()
+            (int row, int column)[] directions = [(0, 1), (1, 0), (1, 1), (1, -1)];
+
+            for (int i = 1; i <= 4; i++)
+            {
+                // The modification of i won't affect these lines.
+                // ReSharper disable once AccessToModifiedClosure
+                var rowChess = ChessPlacements.Where(c => c.BoardRow == i);
+
+                foreach (var chess in rowChess)
+                {
+                    if (chess.CurrentType == ChoiceType.RedWin)
+                    {
+                        foreach (var d in directions)
+                            num.red = Math.Max(num.red, progress(d.row, d.column, ChoiceType.RedWin, chess.BoardRow, chess.BoardColumn));
+                    }
+                    else if (chess.CurrentType == ChoiceType.BlueWin)
+                    {
+                        foreach (var d in directions)
+                            num.blue = Math.Max(num.blue, progress(d.row, d.column, ChoiceType.BlueWin, chess.BoardRow, chess.BoardColumn));
+                    }
+                }
+            }
+
+            if (num.red == 1) num.red = 0;
+            if (num.blue == 1) num.blue = 0;
+
+            return num;
+
+            int progress(int rowDelta, int columnDelta, ChoiceType targetType, int row, int column, int current = 0)
+            {
+                // Step 1: Find the next chess; Return if not found or not desired type (range check implicitly included)
+                var nextChess = ChessPlacements.FirstOrDefault(c => c.BoardRow == row && c.BoardColumn == column);
+
+                if (nextChess == null || nextChess.CurrentType != targetType)
+                    // Edge case: Dismiss dual diagonal matches
+                    return rowDelta == 1 && columnDelta != 0 && current <= 2 ? 0 : current;
+
+                // Step 2: Search forwards
+                return progress(rowDelta, columnDelta, targetType, row + rowDelta, column + columnDelta, ++current);
+            }
+        }
```

Based on my understanding the method `GetMaximumSuccessiveChess` will run a scan using `FirstOrDefault`, works implicitly as a boundary check because if a coordinate is off the board—or simply unoccupied—it won’t be present in ChessPlacements. In theory if no chess are being placed outside of the board then it should be okay.

---

This is based on the assertion that chess can not be placed out of the valid board area, since the placement record only comes from the board. And it should be okay.

If still uncertain, we can make it throw an exception when an invalid placement is made somewhere.

---

Update: Seemingly not rubust enough (external changes considertaion). Holding until resolved.

---

Update 2: Boundary check and relevant tests added.